### ISSUE TITLE
Deploy on multiple Managed Servers

### ISSUE THREAD

**Issue by augeivv** [oracle/weblogic-monitoring-exporter#368]
I think the moitoring exporter is great. However, we have a fundamental problem.
However, we would like to deploy it in a classic weblogic domain in which the administration port is activated. 
According to the documentation, the REST port must then be specified.
The whole thing also works for a single managed server.
However, if I want to deploy the whole thing on all managed servers, the administration port differs per server.
This means that I would have to adjust the configuration several times per server and then also deploy the application several times.
Is there an easier way to do this, where the port can be determined dynamically or something like that?

Thank you very much

**Comment by russgold** [comment]
The REST port only needs to be specified if it is different from the port on which you are contacting the exporter. Under most conditions, it would be the same, so you don't need to specify it.

The most common in which it is not the same is when you have a load balancer in the middle, which will redirect requests to different ports that the one on the load balancer.

**Comment by augeivv** [comment]
Hello @russgold,
is there any news on this topic yet?

**Comment by russgold** [comment]
@augeivv That sounds as though it would need an enhancement to the syntax. Do you have some ideas on what would work? A single configuration would need some way to tell which port is used where.

Do you have an Oracle support person who could look at your application? I don't understand why each server would need its own port numbers.

**Comment by augeivv** [comment]
To the first question:
I could imagine that a query of the ServerRuntime could be built into the Monitoring Exporter.
The Monitoring Exporter should use the Administration URL for the query.

This is available in the ServerRuntime under 'getAdministrationURL()'.
https://docs.oracle.com/middleware/12212/wls/WLAPI/weblogic/management/runtime/ServerRuntimeMBean.html#getAdministrationURL--

When the administration port is activated, the operation returns the URL that the server and its clients use for administrative connections.
If this is not active, it returns the default channel URl.

To the second question:
We would like to use the monitoring exporter for our classic WLS installations as well. So not in the Kubernetes world.
There is a weblogic domain with several managed servers installed on a Linux VM.
Each managed server provides its services on its own port. As two managed servers cannot work on the same port.
In addition, the administration port is active in the domain. This means that the rest api of the weblogic is accessible on an administration channel and not on the default channel.

Furthermore, what exactly do you mean by "your application"?

**Comment by russgold** [comment]
I meant, how you have set up your servers so that each one needs its own port numbers. My understanding was that one usually used the same port numbers on different servers.

**Comment by augeivv** [comment]
Multiple weblogic server instances can also be run on the same server. Nowhere is it defined that each instance must be run on different servers. 
It must only be ensured that they work on different ports so that they can start at all.
I'll look for the exact description in the documentation again. And add it here...
 
In the domain, a separate listen port can be set for each weblogic instance.
The setting can be found in the Administration Console as follows.
expand Environment and select Servers.
On the Servers page, click the name of the server.
Select Configuration > General.

**Comment by russgold** [comment]
The most obvious solution is  not to use the REST_PORT setting and simply access the exporters on the admin ports. Then you would not need a different configuration for each server. Did you try that? If so, what happened?

**Comment by russgold** [comment]
In the usual setup, you would indeed access the exporter via the admin port, if that is enabled.

But I have only worked with two setups - the normal WLS system, in which each server is on a different host, or in Kubernetes with the Weblogic Kubernetes Operator (the easiest choice).

Can you describe your environment? The JVM is a VM, but I am guessing that's not what you mean, here - and you're not using Kubernetes, either?

**Comment by kemalunel** [comment]
Also I'm having an issue like that. I deployed exporter to one of my cluster in wl server but it could not call api to get metrics. it causes, can't see metrics page. I see 404

**Comment by mriccell** [comment]
We will open an enhancement on the monitoring exporter so that the exporter can be accessed via the admin port when serverhave have different admin ports configured (and of course enabled) because the servers are running on the same VM.

**Comment by russgold** [comment]
@kemalunel To be clear, you can reach the main page of the exporter app, but when you hit the `metrics` link, you get a 404? There is also a `messages` link; what happens when you try _that_?

**Comment by kemalunel** [comment]
> @kemalunel To be clear, you can reach the main page of the exporter app, but when you hit the `metrics` link, you get a 404? There is also a `messages` link; what happens when you try _that_? 

Yes I'm able to connect to main page but when I try to upload config file over browser or try to reach metrics and message path, I'm getting the same error 404. 

Those clusters are running on different ports if you compare with admin server. I'm suspecting, it cannot use api feature of weblogic.

**Comment by kemalunel** [comment]
Ok I've fixed 404 error when I try to reach /metrics path with enabling Restful services features on Domain > Advance configuration. but now, when I try to reach metrics page, it asks me username and password.How to deal with that because I'll add this endpoint to prometheus config

**Comment by russgold** [comment]
@augeivv We didn't consider such a configuration when creating the exporter. We're going to look at an enhancement to handle that more easily.
